계산기 II [STEP 1,2]  Kyo, som

KyoPak opened the pull request:
@wonhee009
안녕하세요, 라자냐! Kyo와 som입니다. 반갑습니다!
계산기II Step-1 PR 드립니다..! 잘부탁드립니다!🙇🙇
일단 Kyo의 코드에 som의 코드를 넣어보았습니다 
Kyo 코드의 로직과 som 코드의 로직이 ViewController에서 많이 달랐고, Queue를 구현한 방식도 달라 merge를 하기에는 무리가 있을 것이라고 생각했습니다. 따라서 Kyo 코드에서 som의 코드로 부분부분 정상작동하게끔 refactor를 진행하였습니다!

![스크린샷 2022-10-04 오후 3 26 02](https://user-images.githubusercontent.com/59204352/193749323-5718010a-245b-428e-81cb-0e62dbdcc760.png)


### 진행 단계 : Step1
## ⭐️ 구현한 기능
Kyo의 코드에서 보완해야할 점을 som의 코드로 수정하였고, 서로 이야기 하며 refactor를 진행하였습니다.

⚙️ CalculatorViewController 
  - 기존의 에러메세지를 `Constant` `enum` 내부에 넣고 호출하는 방식으로 변경하였습니다.
  - 계산 결과 20자리 이하로 수정하였습니다.
  - `updateResultLabel()` 을 수정하여 `resultLabel`에도 `,`가 들어갈 수 있게 변경하였습니다.
  - `tappedCaculate()` 메서드의 `do-catch문`을 별도로 분리한 `executeCalculate()`메서드를 구현하여 기능을 분리하였습니다.
  - `makeHistoryStackViewLabel()`과 `updateCalculateHistory()`을 기존의 Kyo 코드와 som의 코드로 병합하여 새로운 코드로 구현해보았습니다.

⚙️ Formula
  - `result()` 메서드에서 사용하는 `rhs` 상수를 `secondNumber`로 네이밍 변경했습니다. 

⚙️ ExpressionParser
  - `parse()` 메서드를 `som`의 코드로 변경했습니다. 

⚙️ Tests
  - `ExpressionParserTests`와 `FormulaTests`에 som의 테스트 케이스가 추가하였습니다.


## ⭐️ 코드를 작성할 때 고민되었던 점
- **리팩토링❓**
계산기I에서 이미 리팩토링이 완료된 서로의 코드를 합치는 것이 어려웠습니다. 서로의 코드에서 어느 부분이 더 적합한 지 판단하는 것이 쉽지 않았습니다. 그리고 앞서 말씀드렸듯이, 이미 리팩토링이 완료된 코드라, 어디를 더 리팩토링을 해야할 지 고민이 많았습니다.  


## ⭐️ 조언을 얻고 싶은 부분 
- **Step 1과 Step 2의 구분**
이번 프로젝트에서 Step 1과 Step 2를 어떻게 구분하면 좋을까요? Step 1에서 서로의 코드를 합치면서 오류가 일어나서 자연스레 리팩토링까지 하게 되었는데, Step 1에서 라자냐의 리뷰를 저희가 또 리팩토링을 하고 merge가 되면 Step 2를 어떻게 진행해야 할지 모르겠습니다... 

- **좋은 로직 선택에 대한 고민**
Kyo와 som의 코드에서 예외 처리를 하는 방법(`throw문`으로 구현할 지, `return문`으로 구현할 지), `StackView`를 만드는 방법, `Expression Parser`의 `parse()` 내부 구현, `Formula`의 `result()` 내부 구현(사용 고차함수 상이, 고차함수 위치 상이) 등등이 달랐는데 좋은 코드를 선택하는 기준을 어떻게 잡으면 좋을지 궁금합니다..!

jsa0224 committed:
feat: Constant 프로퍼티 추가

jsa0224 committed:
refactor: 에러 메세지 -> Constant으로 적용

KyoPak committed:
Merge pull request #1 from KyoPak/Feature/ViewController

refactor: ViewController의 에러 메세지 -> Constant으로 적용

KyoPak committed:
refactor: 계산 결과 20자리 이하로 수정

- 계산 결과 20자리 이하가 되게끔 applyNumberFormatter() 메서드 수정.

jsa0224 committed:
Merge pull request #2 from KyoPak/Feature/ViewController

refactor: 계산 결과 20자리 이하로 수정

jsa0224 committed:
fix: updateResultLabel 메서드 수정

KyoPak committed:
Merge pull request #3 from KyoPak/Feature/ViewController

fix: updateResultLabel 메서드 수정

KyoPak committed:
refactor: setUILabel() 메서드 name 변경 및 setDefaultOperand() 메서드로 중복코드 제거

jsa0224 committed:
Merge pull request #4 from KyoPak/Feature/ViewController

setUILabel() 메서드 name 변경 및 setDefaultOperand() 메서드로 중복코드 제거

jsa0224 committed:
refactor: tappedCalculate 메서드 기능 분리 -> executeCalculate 메서드 생성

KyoPak committed:
Merge pull request #5 from KyoPak/Feature/ViewController

tappedCalculate 메서드 기능 분리 -> executeCalculate 메서드 생성

jsa0224 committed:
refactor: parse 메서드 내부 로직 변경

KyoPak committed:
Merge pull request #6 from KyoPak/Feature/ExpressionParser

parse 메서드 내부 로직 변경

KyoPak committed:
refactor: Formula 내부 반복문에서 rhs 변수명 ➡️ secondNumber 로 변경

- 위의 firstNumber과 name 유사하게끔 secondNumber로 변경

jsa0224 committed:
Merge pull request #7 from KyoPak/Feature/Formula

Formula 내부 반복문에서 rhs 변수명 ➡️ secondNumber 로 naming 변경

jsa0224 committed:
test: 테스트 케이스 추가

KyoPak committed:
Merge pull request #8 from KyoPak/Feature/Tests

테스트 케이스 추가

jsa0224 committed:
refactor: makeHistoryStackViewLabel 메서드, updateCalculateHistory 메서드 내부 코드 수정

jsa0224 committed:
chore: 파일 주석 수정

KyoPak committed:
Merge pull request #9 from KyoPak/Feature/ViewController

makeHistoryStackViewLabel 메서드, updateCalculateHistory 메서드 내부 코드 수정

wonhee009 commented on Calculator/Calculator/Controllers/CalculatorViewController.swift:
```diff
@@ -33,62 +33,51 @@ final class ViewController: UIViewController {
         currentOperand = Constant.defaultZero
     }
-
+    
```
```suggestion

```
필요 없는 스페이스가 들어간거 같습니다 :)

KyoPak replied:
해당부분 확인했습니다! 파일에는 스페이스가 들어가지 않은 것으로 확인됩니다.

wonhee009 commented on Calculator/Calculator/Controllers/CalculatorViewController.swift:
```diff
@@ -33,62 +33,51 @@ final class ViewController: UIViewController {
     
     private func updateResultLabel() {
-        resultLabel.text = currentOperand
+        guard let convertToDouble = Double(currentOperand) else { return }
```
- `operand`여도 충분히 의미 전달이 가능할거 같은데요. 어떻게 생각하시나요?
- `currentOperand`를 Double로 변환하는데 문제가 있으면 return으로 반환될거 같은데요. 해당 케이스는 에러가 아닌걸가요? 만약 해당 케이스도 에러로 봐야한다면 그냥 return을 반환하는게 아니라 에러 핸들링이 필요할거 같습니다.

KyoPak replied:
- 현재 입력한 Operand를 명시적으로 나타내고 싶어서 currentOperand라고 변수명을 설정하였었습니다!
- Double로 변환 실패시 `convertFailToDouble`라는 Error를 throw하게끔 수정하였습니다.

wonhee009 commented on Calculator/Calculator/Controllers/CalculatorViewController.swift:
```diff
@@ -157,20 +145,24 @@ final class ViewController: UIViewController {
-            print("None Operand Error")
+            print(Constant.noneOperand)
         } catch CalculatorError.noneOperator {
-            print("None Operator Error")
+            print(Constant.noneOperator)
         } catch {
-            print("Some Error")
+            print(Constant.error)
```
`Constant.noneOperand`, `Constant.noneOperator`, `Constant.error`는 `Constant.nan`과 다르게 error description 텍스트로 이해했는데요.
error 텍스트들을 `Constant`로 묶기에는 무리가 있지 않나 싶습니다.
[이런](https://developer.apple.com/documentation/foundation/localizederror) 프로토콜을 채택해서 `errorDescription` 프로퍼티를 이용해볼 수 있을거 같습니다.
또한, 에러를 단순히 `print`문을 통해 출력하지 않고, 사용자가 알 수 있도록 alert 등을 이용해보는건 어떨까요?

KyoPak replied:
해당부분 라자냐가 말한대로 Constant로 묶기에는 조금 무리가 있는 것 같아 CalculatorError enum 내부로 위치 변경하였습니다. 그리고 message라는 것을 명시적으로 나타내었습니다.

alert도 고안을 했으나 `resultLabel`에 메시지가 표기되게끔 진행을 하였습니다! 저희가 생각한 방안에 대해서 라자냐의 생각이 궁금합니다! ㅎㅎ

wonhee009 replied:
네넵 좋습니다. 제가 코멘트 남겼던 이유는 alert을 사용해야하는건 아니고, 사용자에게 error를 보여줘야한다고 생각했기 때문에 `resultLabel`에 노출하는 것도 좋은 방법 같습니다 :)

wonhee009 reviewed (changes requested):
수고하셨습니다:)
코멘트 남겨두었으니 확인 부탁드려요 😄 

- Step 1과 Step 2의 구분: 코드를 합치는 과정에서 리팩토링을 진행했지만, 코드를 보다보면 또 리팩토링 혹은 적용하고 싶은 것들이 있을거 같습니다:)
- 좋은 로직 선택에 대한 고민: 제가 생각하는 좋은 로직은 버그가 없어야하고, 같이 작업하는 개발자들끼리 합의가 된 로직여야 하고, 기능이 잘 나눠져 있어야 하고, 다른 개발자가 봤을 때에 의문이 없는 코드여야 한다고 생각합니다.

KyoPak committed:
refactor: Double로 convert실패시 Error throw 하게끔 수정

- Double로 convert실패시 Error throw 하게끔 수정
- 기존 Constant 내부 Error 메세지 프로퍼티를 CalculatorError 내부로 위치 변경 및 Error 메세지 변수명 수정

KyoPak committed:
refactor: Error 처리 시 기존의 단순 print 처리 수정

- Error 처리시 기존의 단순 print처리 방식을 resultLabel에 에러메세지를 표기하는 방식으로 변경

wonhee009 commented on Calculator/Calculator/Controllers/CalculatorViewController.swift:
```diff
@@ -20,11 +20,11 @@ final class ViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
-        setUILabel()
+        setUILabelText()
```
`UI` 없어도 뜻이 통할거 같은데요. 어떻게 생각하실까요?
(Label이 어느정도 UI라는 의미를 내포하고 있을거 같습니다.)

jsa0224 replied:
넵 반영하겠습니다!!

jsa0224 committed:
fix: CalculatorError의 에러 타입 프로퍼티를 message 프로퍼티로 변경

jsa0224 commented on Calculator/Calculator/Common/Error/CalculatorError.swift:
```diff
@@ -8,4 +8,11 @@
     case noneOperator
+    case convertFailToDouble
+    
+    static let noneOperandMessage: String = "None Operand Error"
+    static let noneOperatorMessage: String = "None Operator Error"
+    static let convertFailToDoubleMessage: String = "ConvertFail Error"
+    static let errorMessage: String = "Some Error"
```
- CalculatorError
```swift
var message: String {
        switch self {
        case .noneOperand:
            return "None Operand Error"
        ...
        }
    }
```
- CalculatorViewController
```Swift
catch CalculatorError.noneOperand {
            resultLabel.text = CalculatorError.noneOperand.message
        }
```

라자냐가 조언해주신대로 코드를 수정해도, 만약 Error case를 추가할 일이 생기면 전의 타입 프로퍼티를 새로 생성해야 하는 것처럼 message 프로퍼티에도 case를 추가해야 하는 건 똑같아 보입니다. 그리고 message로 바로 접근하게끔 구현하라는 말씀이 맞으실까요?

wonhee009 replied:
```swift
enum Error {
   case aError
   case bError

   static let aErrorMessage = "something"
   static let bErrorMessage = "something"
}

// 사용처
catch CalculatorError.aError {
   resultLabel.text = CalculatorError.aErrorMessage
} catch CalculatorError.bError {
   resultLabel.text = CalculatorError.bErrorMessage
}
```
처음 코드로는 위와 같이 사용했는데요.
```swift
catch CalculatorError.aError {
   resultLabel.text = CalculatorError.bErrorMessage
}
```
제가 우려했던 상황은 error 타입과 errorMessage가 안 맞는 경우였습니다.
computed property를 제안 드렸던 이유는 error 자체가 message를 갖게 됨으로써 `error.message`의 형태로 접근 가능할거 같아 제안드렸습니다.
```swift
catch {
   if let error = error as? CalculatorError {
      resultLabel.text = error.message
   }
}
```

KyoPak replied:
아!! error를 CalculatorError로 다운캐스팅해서 사용하면 정말 굳이매칭을해주지 않아도 되겠네요..!! 
사실 som이랑 의논했을 때, 어떤 방법을 사용해도 직접 매칭을 해줘야한다고 생각했었는데 저런 방법이 있었네요!! 코드가 훨씬 간결해진 것같습니다.
가르쳐주신 부분은 다른 부분도 응용해보면서 사용하겠습니다 감사합니다 :)

jsa0224 committed:
refactor: setUILabelText() 메서드 명 -> serLabelText()로 변경

KyoPak committed:
refactor: error와 error message 매칭 필요없게 끔 수정 구현

- 다운캐스팅 이용하여 error message처리

wonhee009 commented on Calculator/Calculator/Controllers/CalculatorViewController.swift:
```diff
@@ -48,10 +48,10 @@ final class CalculatorViewController: UIViewController {
-            resultLabel.text = CalculatorError.convertFailToDouble.message
         } catch {
-            resultLabel.text = CalculatorError.someError.message
+            if let error = error as? CalculatorError {
+                resultLabel.text = error.message
+            }
```
```swift
catch {
   let errorMessage: String
   if let error = error as? CalculatorError {
      errorMessage = error.message
   } else {
      errorMessage = CalculatorError.unknown.message
   }

   resultLabel.text = errorMessage
}
```
이렇게 CalculatorError로 캐스팅 되지 않는 에러도 핸들링 할 수 있을거 같습니다 :)

wonhee009 reviewed (approved):
수고하셨습니다 :)
참고할만한 코멘트 남겨두었으니 확인해주세요~
이번 프로젝트는 여기서 마쳐도 좋을거 같습니다!
